### PR TITLE
[IMM32] Restructure! (Retry)

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

The `imm.c` file became big and bloated, so we split the source file.
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Split the `IMM32` code and do formatting.

### CODE REVIEW

**Comment by katahiromz** [comment]
Check time. I will merge 5 days later.

**Comment by katahiromz** [comment]
`candidate.c`: 406 lines.
`guideline.c`: 198 lines.
`ime.c`: 1201 lines.
`imm.c`: 1688 lines.
`keymsg.c`: 688 lines.
`nt3.c`: 203 lines.
`regword.c`: 520 lines.
`utils.c`: 276 lines.

File: dll/win32/imm32/precomp.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include <ddk/imm.h>
+#include <winnls.h>
+#include <winreg.h>
+#include <wine/list.h>
+#include <stdlib.h>
+#include <ndk/umtypes.h>
+#include <ndk/pstypes.h>
+#include <ndk/rtlfuncs.h>
+#include "../../../win32ss/include/ntuser.h"
+#include "../../../win32ss/include/ntwin32.h"
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:35]: Just use
```suggestion
#include "ntuser.h"
#include "ntwin32.h"
```
and in the CMakeFiles.txt above, do:
```
include_directories(
    ${REACTOS_SOURCE_DIR}/sdk/include/reactos/wine
    ${REACTOS_SOURCE_DIR}/win32ss/include)
```

File: dll/win32/imm32/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,7 @@
 
-include_directories(${REACTOS_SOURCE_DIR}/sdk/include/reactos/wine)
+include_directories(
+    ${REACTOS_SOURCE_DIR}/sdk/include/reactos/wine
+    ${REACTOS_SOURCE_DIR}/win32ss/include)
 add_definitions(-D__WINESRC__)
 
 remove_definitions(-D_WIN32_WINNT=0x502)
```
Comment (HBelusca) [inline dll/win32/imm32/CMakeLists.txt:7]:
```suggestion

add_definitions(-D__WINESRC__)
remove_definitions(-D_WIN32_WINNT=0x502)
```

**Review by HBelusca — APPROVED** [review]
minor "formatting" in the cmakefiles, but otherwise LGTM

File: dll/win32/imm32/CMakeLists.txt (hunk 2 above)
Comment (HBelusca) [reply dll/win32/imm32/CMakeLists.txt:7]: by the way, is the `remove_definitions` one necessary?

Comment (katahiromz) [reply dll/win32/imm32/CMakeLists.txt]: When removing `remove_definitions(-D_WIN32_WINNT=0x502)`:
```txt
<command-line>: error: "_WIN32_WINNT" redefined [-Werror]
```

Comment (katahiromz) [reply dll/win32/imm32/CMakeLists.txt]: Both
`remove_definitions(-D_WIN32_WINNT=0x502)` and `add_definitions(-D_WIN32_WINNT=0x600)` removed.

File: dll/win32/imm32/precomp.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+#include <wine/list.h>
+#include <stdlib.h>
+#include <ndk/umtypes.h>
+#include <ndk/pstypes.h>
+#include <ndk/rtlfuncs.h>
+#include "ntuser.h"
+#include "ntwin32.h"
+#include <undocuser.h>
+#include <imm32_undoc.h>
+#include <strsafe.h>
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:38]:
```suggestion
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

/* PSDK/NDK Headers */
#define WIN32_NO_STATUS
#include <windef.h>
#include <winbase.h>
#include <winerror.h>
#include <wingdi.h>
#include <winuser.h>
#include <winnls.h>
#include <winreg.h>

#include <imm.h>
#include <ddk/imm.h>

#define NTOS_MODE_USER
#include <ndk/umtypes.h>
#include <ndk/pstypes.h>
#include <ndk/rtlfuncs.h>

/* Public Win32K Headers */
#include "ntuser.h"
#include "ntwin32.h"

/* Undocumented user definitions */
#include <undocuser.h>
#include <imm32_undoc.h>

#include <strsafe.h>

#include <wine/debug.h>
#include <wine/list.h>
```

Comment (HBelusca) [reply dll/win32/imm32/precomp.h:38]: Are `winerror.h` and `wine/list.h` still useful?

Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: `<winerror.h>` is removed. `<wine/list.h>` is still used from legacy code.

**Review by HBelusca — APPROVED** [review]

**Comment by binarymaster** [comment]
Lol
image

**Pull request merged**